### DISCUSSION TITLE
Installing Node.js

### DISCUSSION THREAD

**Discussion by quackerex** [Homebrew/discussions#4760] [Everyday usage] [answered]
### Output of `brew config`

```shell
NA
```

### Output of `brew doctor`

```shell
NA
```

### Description of issue

It is generally [discouraged to install Python using Homebrew](https://justinmayer.com/posts/homebrew-python-is-not-for-you/). Does this recommendation also apply to installing Node.js on macOS using Homebrew?

**Comment by SMillerDev** [comment, marked as answer]
That's not what the article says, nor was it made by anyone associated with Homebrew.

What it does say is: if you need a stable version of a programming language that you can change whenever you want. Don't use the one from Homebrew. Use a version manager.

**Reply by quackerex** [reply]
> What it does say is: if you need a stable version of a programming language that you can change whenever you want. Don't use the one from Homebrew. Use a version manager.

Is it acceptable to use Python or Node.js installed through Homebrew If I don't need to change versions? 

Could you also clarify what it means by "if you need a stable version of a programming language". I assume Homebrew provide the same stable versions found in Python or Node.js website.

Also based on this statement, it appears that python installed through Homebrew is for Homebrew to use. Similar to macOS system python which is reserve for the system to use:

> Homebrew’s Python is not for you. It exists to serve Homebrew, or more accurately, Homebrew’s other formulae.

**Reply by SMillerDev** [reply]
> Is it acceptable to use Python or Node.js installed through Homebrew If I don't need to change versions?

You can use whatever you want, as long as you keep in mind that the node formula is a moving target that always ships the latest version.

> Could you also clarify what it means by "if you need a stable version of a programming language". I assume Homebrew provide the same stable versions found in Python or Node.js website.

A stable version as in, not moving. The software is generally just as stable as it is upstream. But if you install node 16.1.0 from the website, I don't think it'll automatically update.

> Also based on this statement, it appears that python installed through Homebrew is for Homebrew to use. Similar to macOS system python which is reserve for the system to use:

It is optimised for Homebrew, but just like the system python... it's your system. Do what you want

**Reply by quackerex** [reply]
> It is optimised for Homebrew, but just like the system python... it's your system. Do what you want

Is it intended to be used outside of Homebrew? I don't want to use something if it's not its intended use case. I've also heard that the system Python is deprecated and could be removed at any time, so I don't want to depend on it.

"Scripting language runtimes such as Python, Ruby, and Perl are included in macOS for compatibility with legacy software. Future versions of macOS won’t include scripting language runtimes by default, and might require you to install additional packages. If your software depends on scripting languages, it’s recommended that you bundle the runtime within the app."

https://developer.apple.com/documentation/macos-release-notes/macos-catalina-10_15-release-notes#Scripting-Language-Runtimes

> But if you install node 16.1.0 from the website, I don't think it'll automatically update.

That's what I've been using so far, and it doesn't automatically update. The fact that it doesn't auto-update is an issue for me because Node.js tends to release [important security updates](https://nodejs.org/en/blog/vulnerability/august-2023-security-releases).

I've also noticed that for Node.js, you can install a specific version (node@X), unlike Python. I have a question regarding why only node@X is marked as keg-only. As far as I know, keg-only status occurs because macOS provides a similar package, and it's done to avoid conflicts.

**Reply by SMillerDev** [reply]
> Is it intended to be used outside of Homebrew? I don't want to use something if it's not its intended use case.

It's intended to be used outside Homebrew, otherwise it would be much harder to use. It is however subject to the same caveats that the writer of the article mentions about python. If you are a developer that needs to use exact versions of software to reproduce bugs, it is not the best tool for the job because of its auto updating.

> I've also noticed that for Node.js, you can install a specific version (node@X), unlike Python. I have a question regarding why only node@X is marked as keg-only. As far as I know, keg-only status occurs because macOS provides a similar package, and it's done to avoid conflicts.

There are python versioned formulae just as they exist for node: https://formulae.brew.sh/formula/python@3.11 

Keg-only is indeed meant to avoid conflicts, but you're wrong about it being exclusively conflicts with macOS. Node 16 installs the exact same files as Node 18, just a different version of them. That is why versioned formulae are also keg-only because they conflict with the main version.

**Reply by quackerex** [reply]
> That is why versioned formulae are also keg-only because they conflict with the main version.

If I have both `node` and `node@18` installed, and `node@18` is included in the system's PATH environment variable, could this configuration potentially cause any problems down the road?
I assume Homebrew marks versioned formulae as keg-only to prevent any potential ambiguity. Formulae dependent on others should be capable of correctly selecting the appropriate version, disregarding any custom path configurations in .zshrc.

**Reply by SMillerDev** [reply]
The ambiguity is really the issue. Homebrew tries to ensure tools will pick the right version of dependencies, but we can't really guarantee it for software the team didn't write themselves.
